### ISSUE TITLE
Improvements to file upload security

### ISSUE THREAD

**Issue by kgf** [ReadyResponder/ReadyResponder#429]
The following is a quote from @houhoulis in issue #306. I would like to implement this, but #306 will be closed soon, and I did not want to loose track of this.

> I'm not sure where you saw this behavior, or how many filenames you tested with. But I can see a routing error locally if I use a file with dots in its name, such as "Pic 2017-01-01 at 10.10.10am.png". With the routing error, I see the filename in the browser rather than the picture, as you reported.
> 
> This appears to be Rails behaving as designed. See
> http://guides.rubyonrails.org/routing.html#dynamic-segments
> 
> One fix I found on Stackoverflow is to change the route declaration to allow dots in the image filename:
> 
>   get '/uploads/person/portrait/:id/:basename.:extension', controller: 'people',
>     action: 'download', type: 'portrait',
>     constraints: { basename: /[^\/]+/ }  # <--- this "constraints" part would be new
> But I'd rather handle filenames a bit more carefully, for a couple reasons. One is it feels sort of creepy to show the person's original filename in everyone's browser network activity. Another is I worry we might hit additional edge cases where the filename causes problems in Rails. Including possible hacking attempts, where someone attempts to get ReadyResponder to misinterpret malformed filenames. (The Carrierwave README mentions this briefly.)
> 
> I think Carrierwave might have an option to generate a filename based on the hash of the contents. If we can make use of a Carrierwave feature like that, that feels smarter than dancing around edge cases ourselves. If not, then (from the README) at least we could set the sanitize_regexp = something strict like /[^[:word:]\-]/ (strip out everything from filenames except the letters, numbers, underscores, and dashes)

**Comment by cesc1989** [comment]
Hi @kgf I've got here following Hacktoberfest tags and I'd like to give this issue a shot.

Might have questions I'd let you know.

**Comment by kgf** [comment]
Great, we'd love the help !
No problem with the questions.

Thanks,
Kevin

**Comment by cesc1989** [comment]
Hey @kgf hope you doing fine.

I've implemented a way to fix this issue but before sending a PR I'd like to know if the file name matters. What I mean is that the solution changes the filename to a unique token

```ruby
def filename
  "#{secure_token}.#{file.extension}" if original_filename
end

private

def secure_token
  var = :"@#{mounted_as}_secure_token"
  model.instance_variable_get(var) or model.instance_variable_set(var, SecureRandom.uuid)
end
```

If there's no problem, I'd create a module defining the `secure_token` method and include it in every uploader.

**Comment by kgf** [comment]
I hope that @houhoulis will also take a look here.
I don't see a problem converting it to a token. That could be better so people can't predict file names.
Will there be any issue with existing data/images ?

**Comment by cesc1989** [comment]
> Will there be any issue with existing data/images ?

When I saw this issue it reminded me of this implementation somewhere else (but wasn't sure until testing it myself) and as far as I know, it shouldn't cause trouble with existing images because this change would only work for future uploads, as the name is the only thing stored in the database.

Also, @kgf before sending the PR, I met two situations.

First, I had to comment line `@item_type.item_category = ItemCategory.find(params[:item_category_id])`   in `item_types_controller` because the link to the form has no parameter indicating such value.

Second, the uploads folder was created at root level and not inside `public/` folder. Not sure why is this happening, any clue?
